File: ads/workspace/detector.py
"""
工作空间检测器

负责自动检测当前工作空间的位置。
"""

import os
import json
from pathlib import Path
from typing import Optional


class WorkspaceDetector:
    """工作空间检测器"""

    WORKSPACE_MARKER = ".ads/workspace.json"
    GIT_MARKER = ".git"

    @staticmethod
    def detect() -> Path:
        """
        检测当前工作空间目录。

        检测优先级：
        1. 环境变量 AD_WORKSPACE
        2. 向上查找 .ads/workspace.json
        3. 向上查找 .git 目录（Git 根目录）
        4. 当前工作目录

        Returns:
            工作空间绝对路径
        """
        # 1. 检查环境变量
        env_workspace = os.getenv("AD_WORKSPACE")
        if env_workspace:
            workspace_path = Path(env_workspace)
            if workspace_path.exists():
                return workspace_path.absolute()

        # 2. 向上查找 .ads/workspace.json
        workspace = WorkspaceDetector._find_marker(WorkspaceDetector.WORKSPACE_MARKER)
        if workspace:
            return workspace

        # 3. 向上查找 .git 目录
        git_root = WorkspaceDetector._find_marker(WorkspaceDetector.GIT_MARKER)
        if git_root:
            return git_root

        # 4. 回退到当前目录
        return Path.cwd().absolute()

    @staticmethod
    def _find_marker(marker: str) -> Optional[Path]:
        """
        从当前目录向上查找包含指定标记文件/目录的目录。

        Args:
            marker: 标记文件/目录名（如 ".git" 或 ".ads/workspace.json"）

        Returns:
            包含标记的目录路径，如果未找到返回 None
        """
        current = Path.cwd().absolute()

        # 最多向上查找 10 层
        for _ in range(10):
            marker_path = current / marker
            if marker_path.exists():
                return current

            # 到达文件系统根目录
            parent = current.parent
            if parent == current:
                break

            current = parent

        return None

    @staticmethod
    def get_workspace_db_path(workspace: Optional[Path] = None) -> Path:
        """
        获取工作空间数据库路径。

        Args:
            workspace: 工作空间路径，None 则自动检测

        Returns:
            数据库文件路径 {workspace}/.ads/ads.db

        Raises:
            RuntimeError: 如果工作空间未初始化
        """
        if workspace is None:
            workspace = WorkspaceDetector.detect()

        db_path = workspace / ".ads" / "ads.db"

        # 不自动创建目录，如果不存在则抛出错误
        if not db_path.parent.exists():
            raise RuntimeError(
                f"工作空间未初始化: {workspace}\n"
                f"请先运行 'ads init' 初始化工作空间"
            )

        return db_path

    @staticmethod
    def get_workspace_rules_dir(workspace: Optional[Path] = None) -> Path:
        """
        获取工作空间规则目录。

        Args:
            workspace: 工作空间路径，None 则自动检测

        Returns:
            规则目录路径 {workspace}/.ads/rules/

        Raises:
            RuntimeError: 如果工作空间未初始化
        """
        if workspace is None:
            workspace = WorkspaceDetector.detect()

        rules_dir = workspace / ".ads" / "rules"

        # 不自动创建目录，如果不存在则抛出错误
        if not rules_dir.exists():
            raise RuntimeError(
                f"工作空间未初始化: {workspace}\n"
                f"请先运行 'ads init' 初始化工作空间"
            )

        return rules_dir

    @staticmethod
    def get_workspace_specs_dir(workspace: Optional[Path] = None) -> Path:
        """
        获取工作空间 specs 目录。

        Args:
            workspace: 工作空间路径，None 则自动检测

        Returns:
            Specs 目录路径 {workspace}/docs/specs/

        Raises:
            RuntimeError: 如果工作空间未初始化
        """
        if workspace is None:
            workspace = WorkspaceDetector.detect()

        specs_dir = workspace / "docs" / "specs"

        # 不自动创建目录，如果不存在则抛出错误
        if not specs_dir.exists():
            raise RuntimeError(
                f"工作空间未初始化: {workspace}\n"
                f"请先运行 'ads init' 初始化工作空间"
            )

        return specs_dir

    @staticmethod
    def is_initialized(workspace: Optional[Path] = None) -> bool:
        """
        检查工作空间是否已初始化。

        Args:
            workspace: 工作空间路径，None 则自动检测

        Returns:
            是否已初始化
        """
        if workspace is None:
            workspace = WorkspaceDetector.detect()

        marker_path = workspace / WorkspaceDetector.WORKSPACE_MARKER
        return marker_path.exists()

    @staticmethod
    def initialize(workspace: Optional[Path] = None, name: Optional[str] = None) -> Path:
        """
        初始化工作空间。

        创建：
        - .ads/workspace.json
        - .ads/ads.db (空数据库)
        - .ads/rules/
        - docs/specs/

        Args:
            workspace: 工作空间路径，None 则使用当前目录
            name: 工作空间名称，None 则使用目录名

        Returns:
            工作空间路径
        """
        if workspace is None:
            workspace = Path.cwd().absolute()
        else:
            workspace = Path(workspace).absolute()

        if name is None:
            name = workspace.name

        # 创建目录结构
        ads_dir = workspace / ".ads"
        ads_dir.mkdir(exist_ok=True)

        rules_dir = ads_dir / "rules"
        rules_dir.mkdir(exist_ok=True)

        specs_dir = workspace / "docs" / "specs"
        specs_dir.mkdir(parents=True, exist_ok=True)

        # 创建工作空间配置
        from datetime import datetime

        workspace_config = {
            "name": name,
            "created_at": datetime.now().isoformat(),
            "version": "1.0"
        }

        config_file = ads_dir / "workspace.json"
        with open(config_file, 'w', encoding='utf-8') as f:
            json.dump(workspace_config, f, indent=2, ensure_ascii=False)

        # 初始化数据库
        db_path = ads_dir / "ads.db"
        if not db_path.exists():
            # 创建空数据库（会在首次使用时自动初始化表结构）
            db_path.touch()

        return workspace

    @staticmethod
    def get_workspace_info(workspace: Optional[Path] = None) -> dict:
        """
        获取工作空间信息。

        Args:
            workspace: 工作空间路径，None 则自动检测

        Returns:
            工作空间信息字典
        """
        if workspace is None:
            workspace = WorkspaceDetector.detect()

        config_file = workspace / WorkspaceDetector.WORKSPACE_MARKER

        info = {
            "path": str(workspace),
            "is_initialized": config_file.exists(),
            "db_path": str(WorkspaceDetector.get_workspace_db_path(workspace)),
            "rules_dir": str(WorkspaceDetector.get_workspace_rules_dir(workspace)),
            "specs_dir": str(WorkspaceDetector.get_workspace_specs_dir(workspace))
        }

        # 读取配置
        if config_file.exists():
            try:
                with open(config_file, 'r', encoding='utf-8') as f:
                    config = json.load(f)
                    info.update(config)
            except Exception:
                pass

        return info

File: src/workspace/detector.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const WORKSPACE_MARKER = ".ads/workspace.json";
const GIT_MARKER = ".git";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.resolve(__dirname, "..", "..");
const TEMPLATE_ROOT_DIR = path.join(PROJECT_ROOT, "templates");
const REQUIRED_TEMPLATE_FILES = [
  "instructions.md",
  "rules.md",
  "requirement.md",
  "design.md",
  "implementation.md",
  "workflow.yaml",
];
const LEGACY_TEMPLATE_DIRS = ["nodes", "workflows"];

function existsSync(target: string): boolean {
  try {
    fs.accessSync(target);
    return true;
  } catch {
    return false;
  }
}

function resolveAbsolute(target: string): string {
  return path.resolve(target);
}

function listTemplateFiles(): string[] {
  if (!existsSync(TEMPLATE_ROOT_DIR)) {
    return [];
  }
  const entries = fs.readdirSync(TEMPLATE_ROOT_DIR, { withFileTypes: true });
  const unexpectedDirs = entries.filter((entry) => entry.isDirectory());
  if (unexpectedDirs.length > 0) {
    console.warn(
      `[Workspace] templates/ 目录包含未使用的子目录: ${unexpectedDirs
        .map((entry) => entry.name)
        .join(", ")}`
    );
  }
  const files = entries.filter((entry) => entry.isFile()).map((entry) => entry.name);
  const missing = REQUIRED_TEMPLATE_FILES.filter((file) => !files.includes(file));
  if (missing.length > 0) {
    throw new Error(`templates/ 缺少必需文件: ${missing.join(", ")}`);
  }
  return files;
}

function hasLegacyStructure(dir: string): boolean {
  if (!existsSync(dir)) {
    return false;
  }
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  return entries.some((entry) => entry.isDirectory() || LEGACY_TEMPLATE_DIRS.includes(entry.name));
}

function backupLegacyTemplates(dir: string): void {
  if (!existsSync(dir)) {
    return;
  }
  const backupDir = `${dir}_legacy_${Date.now()}`;
  fs.renameSync(dir, backupDir);
  console.warn(`[Workspace] 发现旧模板结构，已备份到 ${backupDir}`);
}

function filesEqual(a: string, b: string): boolean {
  if (!existsSync(a) || !existsSync(b)) {
    return false;
  }
  const source = fs.readFileSync(a);
  const target = fs.readFileSync(b);
  return source.equals(target);
}

function copyDefaultTemplates(workspaceRoot: string): void {
  const templateFiles = listTemplateFiles();
  if (templateFiles.length === 0) {
    return;
  }

  const templatesRoot = path.join(workspaceRoot, ".ads", "templates");
  if (hasLegacyStructure(templatesRoot)) {
    backupLegacyTemplates(templatesRoot);
  }
  fs.mkdirSync(templatesRoot, { recursive: true });

  const srcSet = new Set(templateFiles);
  for (const entry of fs.readdirSync(templatesRoot)) {
    const entryPath = path.join(templatesRoot, entry);
    const stat = fs.statSync(entryPath);
    if (stat.isDirectory()) {
      fs.rmSync(entryPath, { recursive: true, force: true });
      continue;
    }
    if (!srcSet.has(entry)) {
      fs.rmSync(entryPath, { force: true });
    }
  }

  for (const file of templateFiles) {
    const srcPath = path.join(TEMPLATE_ROOT_DIR, file);
    const destPath = path.join(templatesRoot, file);
    if (filesEqual(srcPath, destPath)) {
      continue;
    }
    fs.copyFileSync(srcPath, destPath);
  }

  const workspaceRulesPath = path.join(workspaceRoot, ".ads", "rules.md");
  const defaultRulesPath = path.join(templatesRoot, "rules.md");
  if (!existsSync(workspaceRulesPath) && existsSync(defaultRulesPath)) {
    fs.copyFileSync(defaultRulesPath, workspaceRulesPath);
  }
}

function findMarker(marker: string, startDir: string, maxDepth = 10): string | null {
  let current = resolveAbsolute(startDir);
  for (let depth = 0; depth < maxDepth; depth += 1) {
    const candidate = path.join(current, marker);
    if (existsSync(candidate)) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }
  return null;
}

export function detectWorkspace(): string {
  const envWorkspace = process.env.AD_WORKSPACE;
  if (envWorkspace && existsSync(envWorkspace)) {
    return resolveAbsolute(envWorkspace);
  }

  const markerDir = findMarker(WORKSPACE_MARKER, process.cwd());
  if (markerDir) {
    return markerDir;
  }

  const gitDir = findMarker(GIT_MARKER, process.cwd());
  if (gitDir) {
    return gitDir;
  }

  return resolveAbsolute(process.cwd());
}

function ensureInitialized<T>(provider: () => T, message: () => string): T {
  try {
    return provider();
  } catch (error) {
    throw new Error(
      `${message()}\n请先运行 'ads init' 初始化工作空间`,
      { cause: error instanceof Error ? error : undefined }
    );
  }
}

export function getWorkspaceDbPath(workspace?: string): string {
  const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();

  // 始终尊重环境变量覆盖（测试场景依赖 ADS_DATABASE_PATH）
  const envDb = process.env.ADS_DATABASE_PATH || process.env.DATABASE_URL;
  if (envDb) {
    const normalized = envDb.replace(/^sqlite:\/\//, "");
    const resolved = path.isAbsolute(normalized) ? normalized : path.resolve(normalized);
    const dir = path.dirname(resolved);
    fs.mkdirSync(dir, { recursive: true });
    if (!existsSync(resolved)) {
      fs.writeFileSync(resolved, "");
    }
    return resolved;
  }

  const dbPath = path.join(root, ".ads", "ads.db");
  if (!existsSync(path.dirname(dbPath))) {
    throw new Error(`工作空间未初始化: ${root}`);
  }
  if (!existsSync(dbPath)) {
    fs.writeFileSync(dbPath, "");
  }
  return dbPath;
}

export function getWorkspaceRulesDir(workspace?: string): string {
  return ensureInitialized(() => {
    const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();
    const rulesDir = path.join(root, ".ads", "rules");
    if (!existsSync(rulesDir)) {
      throw new Error(`规则目录不存在: ${rulesDir}`);
    }
    return rulesDir;
  }, () => `工作空间未初始化: ${workspace ?? detectWorkspace()}`);
}

export function getWorkspaceSpecsDir(workspace?: string): string {
  const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();
  const specDir = path.join(root, "docs", "spec");
  if (existsSync(specDir)) {
    return specDir;
  }

  const legacyDir = path.join(root, "docs", "specs");
  if (existsSync(legacyDir)) {
    try {
      fs.mkdirSync(specDir, { recursive: true });
    } catch (error) {
      throw new Error(`无法创建新的 specs 目录: ${specDir}，原因: ${(error as Error).message}`);
    }
    return specDir;
  }

  try {
    fs.mkdirSync(specDir, { recursive: true });
  } catch (error) {
    throw new Error(`无法创建 specs 目录: ${specDir}，原因: ${(error as Error).message}`);
  }
  return specDir;
}

export function isWorkspaceInitialized(workspace?: string): boolean {
  const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();
  return existsSync(path.join(root, WORKSPACE_MARKER));
}

export function initializeWorkspace(workspace?: string, name?: string): string {
  const root = workspace ? resolveAbsolute(workspace) : resolveAbsolute(process.cwd());
  const workspaceName = name ?? path.basename(root);

  const adsDir = path.join(root, ".ads");
  fs.mkdirSync(adsDir, { recursive: true });

  const specsDir = path.join(root, "docs", "spec");
  fs.mkdirSync(specsDir, { recursive: true });

  const config = {
    name: workspaceName,
    created_at: new Date().toISOString(),
    version: "1.0",
  };

  fs.writeFileSync(
    path.join(adsDir, "workspace.json"),
    JSON.stringify(config, null, 2),
    "utf-8"
  );

  const dbPath = path.join(adsDir, "ads.db");
  if (!existsSync(dbPath)) {
    fs.writeFileSync(dbPath, "");
  }

  copyDefaultTemplates(root);

  return root;
}

export function ensureDefaultTemplates(workspace?: string): void {
  const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();
  copyDefaultTemplates(root);
}

export function getWorkspaceInfo(workspace?: string): Record<string, unknown> {
  const root = workspace ? resolveAbsolute(workspace) : detectWorkspace();
  const configFile = path.join(root, WORKSPACE_MARKER);

  const info: Record<string, unknown> = {
    path: root,
    is_initialized: existsSync(configFile),
    db_path: getWorkspaceDbPath(root),
  };

  try {
    info.rules_dir = getWorkspaceRulesDir(root);
  } catch {
    info.rules_dir = null;
  }

  try {
    info.specs_dir = getWorkspaceSpecsDir(root);
  } catch {
    info.specs_dir = null;
  }

  if (existsSync(configFile)) {
    try {
      const configContent = fs.readFileSync(configFile, "utf-8");
      const parsed = JSON.parse(configContent);
      Object.assign(info, parsed);
    } catch {
      // ignore malformed config
    }
  }

  return info;
}
